mountpoint-s3-fs/src/manifest.rs

```diff
@@ -14,4 +20,138 @@ pub enum ManifestError {
     InvalidKey(String),
+    #[error("invalid database row")]
+    InvalidRow,
+}
+
+/// An entry returned by manifest_lookup() and ManifestIter::next()
+#[derive(Debug, Clone)]
+pub enum ManifestEntry {
+    File {
+        full_key: String,
+        etag: String,
+        size: usize,
+    },
+    Directory {
+        full_key: String, // doesn't contain '/'
+    },
+}
+
+impl TryFrom<DbEntry> for ManifestEntry {
+    type Error = ManifestError;
+
+    fn try_from(db_entry: DbEntry) -> Result<Self, Self::Error> {
+        if db_entry.full_key.ends_with('/') {
+            Err(ManifestError::InvalidRow)
+        } else if db_entry.etag.is_none() && db_entry.size.is_none() {
+            Ok(ManifestEntry::Directory {
+                full_key: db_entry.full_key,
+            })
+        } else if db_entry.etag.is_some() && db_entry.size.is_some() {
+            Ok(ManifestEntry::File {
+                full_key: db_entry.full_key,
+                etag: db_entry.etag.unwrap(),
+                size: db_entry.size.unwrap(),
+            })
+        } else {
+            Err(ManifestError::InvalidRow)
+        }
+    }
+}
+
+/// Manifest of all available objects in the bucket
+#[derive(Debug, Clone)]
+pub struct Manifest {
+    db: Db,
+}
+
+impl Manifest {
+    pub fn new(manifest_db_path: &Path) -> Result<Self, rusqlite::Error> {
+        let db = Db::new(manifest_db_path)?;
+        Ok(Self { db })
+    }
+
+    /// Lookup an entry in the manifest, the result may be a file or a directory
+    pub fn manifest_lookup(
+        &self,
+        parent_full_path: String,
+        name: &str,
+    ) -> Result<Option<ManifestEntry>, ManifestError> {
+        trace!("using manifest to lookup {} in {}", name, parent_full_path);
+        let mut full_path = parent_full_path;
+        full_path.push_str(name);
+
+        // search for an entry and validate it
+        let db_entry = self.db.select_entries(&full_path)?;
+        let Some(db_entry) = db_entry else {
+            return Ok(None);
+        };
+        let manifest_entry = ManifestEntry::try_from(db_entry)?;
+        Ok(Some(manifest_entry))
+    }
+
+    /// Create an iterator over directory's direct children
+    pub fn iter(&self, bucket: &str, directory_full_path: &str) -> Result<ManifestIter, InodeError> {
+        ManifestIter::new(self.db.clone(), bucket, directory_full_path)
+    }
+}
+
+#[derive(Debug)]
+pub struct ManifestIter {
+    db: Db,
+    /// Prepared entries in order to be returned by the iterator.
+    entries: VecDeque<DbEntry>,
+    /// Key of the directory being listed by this iterator
+    parent_key: String,
+    /// Offset of the next child to search for in the database
+    next_offset: usize,
+    /// Max amount of entries to read from the database at once
+    batch_size: usize,
+    /// Database has no more entries
+    finished: bool,
+}
+
+impl ManifestIter {
+    fn new(db: Db, _bucket: &str, parent_key: &str) -> Result<Self, InodeError> {
```

Comment:
Why does this return a result if we can never return an error here?

Reply:
good point, fixed (previously it made a query to db here)